Source/Core/DolphinQt2/Settings/GameCubePane.cpp

```diff
@@ -0,0 +1,333 @@
+  connect(m_language_combo, static_cast<void (QComboBox::*)(int)>(&QComboBox::currentIndexChanged),
+          this, &GameCubePane::SaveSettings);
+  connect(m_override_language_ntsc, &QCheckBox::stateChanged, this, &GameCubePane::SaveSettings);
+
+  // Device Settings
+  for (int i = 0; i < 3; i++)
+  {
+    connect(m_slot_combos[i],
+            static_cast<void (QComboBox::*)(int)>(&QComboBox::currentIndexChanged), this,
+            &GameCubePane::SaveSettings);
+    if (i < 2)
+    {
+      connect(m_slot_buttons[i], &QPushButton::pressed, this, [this, i] { OnConfigPressed(i); });
+    }
+  }
+}
+
+void GameCubePane::OnConfigPressed(int slot)
+{
+  QString filter;
+  bool memcard = false;
+
+  switch (m_slot_combos[slot]->currentIndex())
+  {
+  // Memory card
+  case 2:
+    filter = tr("GameCube Memory Cards (*.raw,*.gcp)");
+    memcard = true;
+    break;
+  // Advance Game Port
+  case 5:
+    filter = tr("Game Boy Advance Carts (*.gba)");
+    memcard = false;
+    break;
+  // Microphone
+  case 6:
+    MappingWindow(this, MappingWindow::Type::MAPPING_GC_MICROPHONE, slot).exec();
+    return;
+  }
+
+  QString filename =
+      QFileDialog::getOpenFileName(this, tr("Choose a file to open"), QStringLiteral(""), filter);
+
+  if (filename.isEmpty() || !File::Exists(filename.toStdString()))
+    return;
+
+  QString path_abs = QFileInfo(filename).absolutePath();
+
+  // Memcard validity checks
+  if (memcard)
+  {
+    GCMemcard mc(filename.toStdString());
+
+    if (!mc.IsValid())
+    {
+      QMessageBox::critical(this, tr("Error"), tr("Cannot use that file as a memory card.\n%s\n"
+                                                  "is not a valid GameCube memory card file")
+                                                   .arg(filename));
+      return;
+    }
+
+    QString path_b =
+        QFileInfo(QString::fromStdString(slot == 0 ? SConfig::GetInstance().m_strMemoryCardB :
+                                                     SConfig::GetInstance().m_strMemoryCardA))
+            .absolutePath();
+
+    if (path_abs == path_b)
+    {
+      QMessageBox::critical(this, tr("Error"),
+                            tr("Are you trying to use the same file in both slots?"));
+      return;
+    }
+  }
+
+  if (memcard)
+  {
+    if (slot == 0)
+    {
+      SConfig::GetInstance().m_strMemoryCardA = path_abs.toStdString();
+    }
+    else
+    {
+      SConfig::GetInstance().m_strMemoryCardB = path_abs.toStdString();
+    }
+  }
+  else
+  {
+    if (slot == 0)
+    {
+      SConfig::GetInstance().m_strGbaCartA = path_abs.toStdString();
+    }
+    else
+    {
+      SConfig::GetInstance().m_strGbaCartB = path_abs.toStdString();
+    }
+  }
+}
+
+void GameCubePane::LoadSettings()
+{
+  const SConfig& params = SConfig::GetInstance();
+
+  // IPL Settings
+  m_skip_main_menu->setChecked(params.bHLE_BS2);
+  m_language_combo->setCurrentIndex(params.SelectedLanguage);
```

Comment:
Is this safe? Can `SConfig::SelectedLanguage` have an out-of-bounds value?

Reply:
It can have an out of bounds value if the INI has been edited by e.g. the user.

Reply:
I'm not sure whether this is something worth addressing. AFAIK Wx doesn't do any sanity checking on any values either,